compiler/src/iree/compiler/Codegen/Common/GPU/GPUNestedLayoutDistributionPatterns.cpp

```diff
@@ -498,6 +510,170 @@ struct DistributeTransferWrite final
+    auto vectorType = VectorType::get(distShape, elementType);
+    // The shape of the vector we read is pre-permutation. The permutation is
+    // a transpose on the resulting read vector.
+    auto innerVectorType =
+        VectorType::get(vectorLayout.getElementTile(), elementType);
+
+    // Initialize the full distributed vector for unrolling the batch/outer
+    // vector dimensions.
+    Value zero = rewriter.create<arith::ConstantOp>(
+        gatherOp.getLoc(), vectorType, rewriter.getZeroAttr(vectorType));
+    VectorValue acc = cast<VectorValue>(zero);
+
+    SmallVector<Value> warpIndices, threadIndices;
+    if (failed(populateWarpAndThreadIndices(rewriter, threadId, subgroupSize,
+                                            vectorLayout, warpIndices,
+                                            threadIndices))) {
+      return rewriter.notifyMatchFailure(
+          gatherOp, "warp or thread tiles have overlapping strides");
+    }
+
+    ValueRange indices = gatherOp.getIndices();
+    AffineMap permMap = gatherOp.getPermutationMap();
+    SmallVector<int64_t> strides(rank, 1);
+
+    SmallVector<SmallVector<int64_t>> allMaskOffsets;
+    std::vector<StaticTileOffsetRange::IteratorTy> allIndexVecOffsets;
+    if (mask) {
+      SmallVector<int64_t> maskDistShape = maskLayout.getDistributedShape();
+      SmallVector<int64_t> maskTileShape =
+          getElementVectorTileShape(maskLayout);
+      allMaskOffsets =
+          llvm::to_vector(StaticTileOffsetRange(maskDistShape, maskTileShape));
+    }
+    for (NestedLayoutAttr layout : indexVecLayouts) {
+      SmallVector<int64_t> vecDistShape = layout.getDistributedShape();
+      SmallVector<int64_t> vecTileShape = getElementVectorTileShape(layout);
+      allIndexVecOffsets.push_back(
+          StaticTileOffsetRange(vecDistShape, vecTileShape).begin());
+    }
+
+    auto indexed =
+        llvm::to_vector(gatherOp.getIndexed().getAsValueRange<BoolAttr>());
+
+    for (auto [idx, offsets] :
+         llvm::enumerate(StaticTileOffsetRange(distShape, tileShape))) {
+      SmallVector<Value> slicedIndices = getTransferIndicesFromNestedLayout(
+          rewriter, indices, offsets, vectorLayout, permMap, warpIndices,
+          threadIndices);
+
+      // Only take the sliced indices for the non-indexed values.
+      for (auto [dim, slicedIndex, index] :
+           llvm::enumerate(slicedIndices, indices)) {
+        if (indexed[dim]) {
+          slicedIndex = index;
+        }
+      }
+
+      // Extract offset from index_vecs.
+      SmallVector<Value> slicedIndexVecs;
+      for (auto [indexVecIdx, disIndexVec, layout] :
+           llvm::enumerate(disIndexVecs, indexVecLayouts)) {
+        auto offsets = llvm::to_vector(*(allIndexVecOffsets[indexVecIdx]));
+        ++allIndexVecOffsets[indexVecIdx];
+        VectorValue slicedIndexVec = getSlicedIndexVec(
+            rewriter, gatherOp.getLoc(), offsets, layout, disIndexVec);
```

Comment:
I think you can construct an arrayref from a range and skip the smallvector construction. Idk exactly what `StaticTileOffsetRange::IteratorTy` is though.
```suggestion
            rewriter, gatherOp.getLoc(), ArrayRef(allIndexVecOffsets[indexVecIdx]->begin(), allIndexVecOffsets[indexVecIdx]->end()), layout, disIndexVec);
```

Reply:
I tried it but i dont think its possible. The operator * for that iterator returns a SmallVector that is being constructed here.

Reply:
aw ok, thanks for trying then.